# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/nsAccUtils.cpp
# Commit: ae4f46dfec58
# Full Hash: ae4f46dfec5886a64797be7073f63bda296c3541
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2025-09-29 20:54 +0000
# Regressor Bug: 1987945
# File Overlap Count: 2
# Overlapping Files: accessible/base/nsCoreUtils.cpp, accessible/tests/browser/relations/browser_anchor_positioning.js
# Description:
#   Bug 1987945 - P1: Implement details relationship for CSS anchored local accessibles. r=Jamie,emilio
# ==============================================================================

diff -r 55449173b0e6 -r ae4f46dfec58 accessible/base/nsAccUtils.cpp
--- a/accessible/base/nsAccUtils.cpp	Mon Sep 29 20:39:23 2025 +0000
+++ b/accessible/base/nsAccUtils.cpp	Mon Sep 29 20:54:29 2025 +0000
@@ -651,3 +651,29 @@
   return aAccessible->IsTextField() ||
          aAccessible->Elm()->State().HasState(dom::ElementState::READWRITE);
 }
+
+bool nsAccUtils::IsValidDetailsTargetForAnchor(const Accessible* aTarget,
+                                               const Accessible* aAnchor) {
+  if (aAnchor->IsAncestorOf(aTarget)) {
+    // If the anchor is a parent of the target, the target is not valid
+    // relation.
+    return false;
+  }
+
+  Accessible* nextSibling = aAnchor->NextSibling();
+  if (nextSibling && nextSibling->IsTextLeaf()) {
+    nsAutoString text;
+    nextSibling->Name(text);
+    if (nsCoreUtils::IsWhitespaceString(text)) {
+      nextSibling = nextSibling->NextSibling();
+    }
+  }
+
+  if (nextSibling == aTarget) {
+    // If the target is the next sibling of the anchor (ignoring whitespace
+    // text nodes), the target is not a valid relation.
+    return false;
+  }
+
+  return true;
+}